src/lashup_kv.erl

```diff
@@ -470,3 +534,72 @@ kv2map(New, [Old | _]) ->
   Map = kv2map(New),
   #{value := OldValue} = kv2map(Old),
   Map#{old_value => OldValue}.
+
+-spec(op_write(tuple()) -> ok).
+op_write(Record) ->
+  Table = element(1, Record),
+  prometheus_summary:observe_duration(
+    lashup, kv_backend_write_seconds, [Table],
+    fun () -> mnesia:write(Record) end).
+
+%%%===================================================================
+%%% Metrics functions
+%%%===================================================================
+
+-spec(init_metrics() -> ok).
+init_metrics() ->
+  init_kv_metrics(),
+  init_op_metrics(),
+  init_backend_metrics().
+
+-spec(init_kv_metrics() -> ok).
+init_kv_metrics() ->
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_handle_op_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent processing KV operations."}
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_handle_full_update_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent processing KV full updates."}
+  ]),
+  prometheus_gauge:new([
+    {registry, lashup},
+    {name, kv_message_queue},
+    {help, "The length of KV process message box."}
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_crdt_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent merging/updating CRDT in KV process."}
+  ]).
+
+-spec(init_op_metrics() -> ok).
+init_op_metrics() ->
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_request_op_seconds},
+    {duration_unit, seconds},
+    {help, "The time spent waiting for processing KV operations."}
+  ]).
+
+-spec(init_backend_metrics() -> ok).
+init_backend_metrics() ->
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_backend_read_seconds},
+    {labels, [table]},
+    {duration_unit, seconds},
+    {help, "The time spent reading data from KV backend."}
+  ]),
+  prometheus_summary:new([
+    {registry, lashup},
+    {name, kv_backend_write_seconds},
+    {labels, [table]},
```

Comment:
It would be good to also capture information about `Key` as well

Reply:
Prometheus docs recommend not to add such kind of labels [1]. We can add key-specific metrics in place where we use them in dcos-net.

[1] https://prometheus.io/docs/practices/naming/#labels

Reply:
What about the cardinality of tables in lashup? I think I only saw two tables on a quick test I did, but I wonder how many you have on a cluster, and if the size of the cluster matters for the number of tables.

Reply:
There are only two mnesia tables in lashup: `kv` stores `riak_dt_map`s, `nclock` stores metadata for AAE.